File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

bool isValidCardLength(int);
string getCardType(long);
int getNumLength(long);
bool checksum(long);

int main(void)
{
    long cardNumber = get_long("Number: ");
    int length = getNumLength(cardNumber);

    // Check card number length and run checksum algorithm
    if (!isValidCardLength(length) || !checksum(cardNumber))
    {
        printf("INVALID\n");
        return 0;
    }

    printf("%s\n", getCardType(cardNumber));

}

int getNumLength(long num)
{
    return floor(log10(num)) + 1;
}

// American Express uses 15-digit numbers, MasterCard uses 16-digit numbers, and Visa uses 13- and 16-digit numbers
bool isValidCardLength(int length)
{
    return length == 13 || length == 15 || length == 16;
}

// All American Express numbers start with 34 or 37; most MasterCard numbers start with 51, 52, 53, 54, or 55 and all Visa numbers start with 4.
string getCardType(long number)
{
    long num = number;
    int len = getNumLength(number);
    int firstTwo = 0;
    while (--len)
    {
        num /= 10;
        if (len == 2)
        {
            firstTwo = num;
            break;
        }
    }

    if (firstTwo == 34 || firstTwo == 37)
    {
        return "AMEX";
    }
    if (firstTwo >= 51 && firstTwo <= 55)
    {
        return "MASTERCARD";
    }
    if (firstTwo >= 40 && firstTwo <= 49)
    {
        return "VISA";
    }

    return "INVALID";
}


// Luhn's algorithm
// 1. Multiply every other digit by 2, starting with the number’s second-to-last digit, and then add those products’ digits together.
// 2. Add the sum to the sum of the digits that weren’t multiplied by 2.
// 3. If the total’s last digit is 0 (or, put more formally, if the total modulo 10 is congruent to 0), the number is valid!
bool checksum(long number)
{
    long num = number;
    int len = getNumLength(number);
    int i = 0;
    int sum = 0;

    while (len--)
    {
        int digit = num % 10;

        // For every other digit, multiply it by 2
        if (i % 2 != 0)
        {
            int prod = digit * 2;
            if (getNumLength(prod) > 1)
            {
                // Sum both digits of the product, in case it's more than 1
                sum += prod % 10;
                prod /= 10;
                sum += prod % 10;
            }
            else
            {
                sum += prod;
            }
        }
        else
        {
            sum += digit;
        }

        num /= 10;
        i++;
    }

    return sum % 10 == 0;
}

File: pset6/credit/credit.py
from cs50 import get_int


def main():
    # Get user's input and convert it to string, for better handling
    cardNumber = str(get_int("Number: "))

    # Check card number length and run checksum algorithm
    if (not isValidCardLength(cardNumber)) or (not checksum(cardNumber)):
        print("INVALID")
    else:
        print(getCardType(cardNumber))


# American Express uses 15-digit numbers, MasterCard uses 16-digit numbers, and Visa uses 13- and 16-digit numbers
def isValidCardLength(num):
    length = len(num)
    return length == 13 or length == 15 or length == 16


# All American Express numbers start with 34 or 37;
# most MasterCard numbers start with 51, 52, 53, 54, or 55 and
# all Visa numbers start with 4.
def getCardType(num):
    firstTwo = int(num[:2])

    if firstTwo == 34 or firstTwo == 37:
        return "AMEX"
    if 51 <= firstTwo <= 55:
        return "MASTERCARD"
    if 40 <= firstTwo <= 49:
        return "VISA"

    return "INVALID"


# Luhn's algorithm
# 1. Multiply every other digit by 2, starting with the number’s second-to-last digit, and then add those products’ digits together.
# 2. Add the sum to the sum of the digits that weren’t multiplied by 2.
# 3. If the total’s last digit is 0 (or, put more formally, if the total modulo 10 is congruent to 0), the number is valid!
def checksum(num):
    reversedNum = num[::-1]
    total = 0

    for i in range(len(reversedNum)):
        # Check if current digit is on an odd index
        if i % 2 != 0:
            prod = int(reversedNum[i]) * 2

            # add product's digits together, if more than one digit
            # otherwise, only add the product to the total
            digits = str(prod)
            total += sum([int(d) for d in digits])

        else:
            # if digit is on even index, just add it to the total
            total += int(reversedNum[i])

    return total % 10 == 0


# Run main function
main()